nixos/forgejo: change `cfg.database.createDatabase` default from `true` to `false`

Lindenk opened the pull request:
The current Forgejo module attempts to set up a database on the local system by default, which can lead to broken database installations with no warning if the user has Patroni or similar configured. This is both unexpected and dangerous, potentially leading to a borked install and data loss.

In cases where the database host is set to a remote server, the current module still attempts to deploy and configure a database server on the local machine, even though it will be unused. This patch disables this silent behavior

## Things done

Set the default `createDatabase` option to false so the user must opt into hosting a local database

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

NyCodeGHG commented:
You are supposed to set `createDatabase` to false if you configure your database manually.
I think `createDatabase` defaulting to true is the the better option for most users.

Lindenk commented:
Does any other module do this? I would argue it's bad practice for a module to enable and configure another module silently by default

It's also still incorrect behavior if `host` is set to anything remote

The main issue here is that it's not only running sql, but it's also enabling postgres on the system forgejo is running on. It would be much more intuitive if it only ran it's SQL against the configured `host` and didn't assume the user wants to run a database on the same system

Lastly if another system (such as patroni) is enabled, this default will clobber the database on the current system, meaning any patroni clusters using the system as a node will cause the node to fail and break filesystem permissions (or worse). 

Since nixos documentation is scarce (and this is clearly undocumented), leaving in a footgun by default like this seems like a awful design choice

adamcstephens reviewed (changes requested):
We can't change the default like this. If we want to move forward with a different default, we'll have to use `system.stateVersion` to only apply it moving forward, and not cause breakage for existing users using the existing default.

Lindenk commented:
> I also run a patroni cluster, and examine every module I go to enable to understand how it will behave. I think as a user of a more advanced setup, I expect to have to do more effort.

I don't think forcing a user to read through the entire source of every module they plan on enabling is a fair expectation for any configuration. It should be expected that a module handles itself and no more unless explicitly configured

> > argue it's bad practice for a module to enable and configure another module silently by default
> 
> This is done across the module system for all kinds of reasons. Different authors and maintainers have different goals and expectations. I don't always agree with it, but I do agree that out of box experience is important.

A good out of the box is important, but the trade off here is the default is incorrect and dangerous in any use case where the database is remote, which is an awful out of the box experience 

What would you expect this configuration to do:
```nix
services.forgejo = {
  enable = true;
  database = {
    type = "postgres";
    host = "db.example.com";
    user = "forgejo";
    passwordFile = "...";
  };
}
```

This will currently set up forgojo with connection details for a remote postgres database on "db.example.com". It will _also_ set up a database on the system running forgejo which, best case, is unused, or worst case clobbers an existing database. This is unexpected

Lindenk committed:
Forgejo: Changed default database creation strategy to false

The current default is to enable the postgres module and create a user and database for forgejo. This can cause breaking changes to the host's system if another database management system is used (such as Patroni) and is unnecessary when a remote host is defined.

Lindenk commented:
I agree, it would make sense to at least produce a warning if host is set to a remote address and createDatabase is also enabled, although I'm not sure how we'd determine whether or not the host is the same machine. In any case I have it changed for stateVersions >= 25.11
